### PR TITLE
 improve postgres extended query protocol troubleshooting

### PR DESCRIPTION
**Pull request by djova**
### What does this PR do?

* display per-client instructions in a table in a dedicated section
* add instructions for several more well known postgres clients

### Motivation

Improve troubleshooting instructions.

### Additional Notes

---

### Reviewer checklist
- [ ] Review the changed files.
- [ ] Review the URLs listed in the [Preview](#preview) section.
- [ ] Check images for PII
- [ ] Review any mentions of "Contact Datadog support" for internal support documentation.

### CODE REVIEW

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 1]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+|----------|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol.                                                                                                                                                                                                                                                                                                                           |
```
Comment (justiniso) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:188]:
```suggestion
| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol ([Docs](https://pkg.go.dev/github.com/jackc/pgx#ConnConfig)).                                                                                                                                                                                                                                                                                                                           |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 2]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+| Java     | [Postgres JDBC Client][14] | Set `preferQueryMode = simple` to switch to the simple query protocol.                                                                                                                                                                                                                                                                                                                       |
```
Comment (justiniso) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:189]:
```suggestion
| Java     | [Postgres JDBC Client][14] | Set `preferQueryMode = simple` to switch to the simple query protocol ([Docs](https://jdbc.postgresql.org/documentation/publicapi/org/postgresql/jdbc/PreferQueryMode.html)).                                                                                                                                                                                                                                                                                                                       |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 3]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+| Java     | [Postgres JDBC Client][15] | Set `preferQueryMode = simple` to switch to the simple query protocol ([Docs][16]).                                                                                                                                                                                                                                                                                                          |
+| Python   | [asyncpg][17]              | Uses the extended query protocol and it's not possible to disable it. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |
```
Comment (alexandre-normand) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md]: Is there a reason why we don't have the pre-formatted statement workaround for `asyncpg` like we have for the others below? 
```
Format SQL Queries using [psycopg sql][17] before passing them to the DB client to enable collection of execution plans.
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 4]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+|----------|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol ([Docs][14]).                                                                                                                                                                                                                                                                                                              |
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:188]:
```suggestion
| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol (See the [ConnConfig documentation][14]).                                                                                                                                                                                                                                                                                                              |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 5]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+| Java     | [Postgres JDBC Client][15] | Set `preferQueryMode = simple` to switch to the simple query protocol ([Docs][16]).                                                                                                                                                                                                                                                                                                          |
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:189]:
```suggestion
| Java     | [Postgres JDBC Client][15] | Set `preferQueryMode = simple` to switch to the simple query protocol (See the [PreferQueryMode documentation][16]).                                                                                                                                                                                                                                                                                                          |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md (hunk 3 above)
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:190]:
```suggestion
| Python   | [asyncpg][17]              | Uses the extended query protocol, which cannot be disabled. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 6]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+| Python   | [asyncpg][17]              | Uses the extended query protocol and it's not possible to disable it. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |
+| Python   | [psycopg][18]             | `psycopg2` does not use the extended query protocol so execution plans should be collected without issue. <br/> `psycopg3` uses the extended query protocol by default and there is no way to disable it. Disabling prepared statements does not solve the problem. Format SQL Queries using [psycopg sql][19] before passing them to the DB client to enable collection of execution plans. |
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:191]:
```suggestion
| Python   | [psycopg][18]             | `psycopg2` does not use the extended query protocol so execution plans should be collected without issue. <br/> `psycopg3` uses the extended query protocol by default and cannot be disabled. Disabling prepared statements does not solve the problem. To enable the collection of execution plans, format SQL Queries using [psycopg sql][19] before passing them to the DB client. |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 7]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
+| Node     | [node-postgres][20]       | Uses the extended query protocol and it's not possible to disable it. Use [pg-format][21] to format SQL Queries before passing them to [node-postgres][20] to enable the Datadog Agent to collect execution plans.                                                                                                                                                                           |
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:192]:
```suggestion
| Node     | [node-postgres][20]       | Uses the extended query protocol and cannot be disabled. To enable the Datadog Agent to collect execution plans, use [pg-format][21] to format SQL Queries before passing them to [node-postgres][20].                                                                                                                                                                           |
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 8]
```diff
@@ -206,21 +215,27 @@ This error happens when you are missing the `postgresql-contrib` package that in
 sudo apt-get install postgresql-contrib-10
 ```
 
-See the appropriate version of the [Postgres `contrib` documentation][16] for more information.
+See the appropriate version of the [Postgres `contrib` documentation][22] for more information.
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:218]:
```suggestion
For more information, see the appropriate version of the [Postgres `contrib` documentation][22].
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 9]
```diff
@@ -179,17 +179,26 @@ Ensure that the Agent is running version 7.32.0 or newer. Datadog recommends reg
...
 
-#### Application using Postgres extended query protocol or prepared statements
-The application client used to execute the query is using the Postgres extended query protocol or prepared statements. Some client applications using the Postgres [extended query protocol][5] do not support the collection of explain plans due to the separation of the parsed query and raw bind parameters. For instance, the Python client [asyncpg][6] and the Go client [pgx][7] use the extended query protocol by default. To work around this limitation, you can configure your database client to use the simple query protocol. For example: set `preferQueryMode = simple` for the [Postgres JDBC Client][8] or set `PreferSimpleProtocol` on the [pgx][7] connection config.
+#### Postgres extended query protocol
+
+If a client is using the Postgres [extended query protocol][13] or prepared statements then the Datadog Agent is not able to collect explain plans due to the separation of the parsed query and raw bind parameters. If the client provides an option to force using the simple query protocol then turning that on will enable the Datadog Agent to collect execution plans.
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:184]:
```suggestion
If a client is using the Postgres [extended query protocol][13] or prepared statements, the Datadog Agent is unable to collect explain plans due to the separation of the parsed query and raw bind parameters. If the client provides an option to force using the simple query protocol, then turning that on enables the Datadog Agent to collect execution plans.
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 10]
```diff
@@ -144,17 +144,17 @@ The resulting normalized query will appear in the app as:
...
+To avoid this, raise the `track_activity_query_size` setting to a value large enough to accommodate the largest expected text size of your queries. For further information, see the Postgres documentation on [runtime statistics][12].
 
 ### Queries are missing explain plans
 
 Some or all queries may not have plans available. This can be due to unsupported query commands, queries made by unsupported client applications, an outdated Agent, or incomplete database setup. Below are possible causes for missing explain plans.
 
 #### Missing explain function {#undefined-explain-function}
 
-Problem: The Agent is not able to execute a required function in the `datadog` schema of the database. 
+Problem: The Agent is not able to execute a required function in the `datadog` schema of the database.
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:155]:
```suggestion
Problem: The Agent is unable to execute a required function in the `datadog` schema of the database.
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 11]
```diff
@@ -144,17 +144,17 @@ The resulting normalized query will appear in the app as:
...
 
-Solution: The Agent requires the function `datadog.explain_statement(...)` to exist in **all databases** the Agent can collect queries from. 
+Solution: The Agent requires the function `datadog.explain_statement(...)` to exist in **all databases** the Agent can collect queries from.
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:157]:
```suggestion
Solution: The Agent requires the `datadog.explain_statement(...)` function to exist in **all databases** the Agent can collect queries from.
```

File: content/en/database_monitoring/setup_postgres/troubleshooting.md [hunk 12]
```diff
@@ -84,16 +84,16 @@ DD_LOG_LEVEL=debug DBM_THREADED_JOB_RUN_SYNC=true agent check sqlserver -t 2
...
+| Queries are truncated. | See the section on [truncated query samples](#query-samples-are-truncated) for instructions on how to increase the size of sample query text.                                                                                                                                                                                                                                                                                                                              |
+| The application client used to execute the query is using the Postgres extended query protocol or prepared statements. | See the section on [Postgres extended query protocol](#postgres-extended-query-protocol).                                                                                                                                                                                                                                                                                                                                                                                  |
```
Comment (alai97) [inline content/en/database_monitoring/setup_postgres/troubleshooting.md:92]:
```suggestion
| The application client used to execute the query is using the Postgres extended query protocol or prepared statements. | See [Postgres extended query protocol](#postgres-extended-query-protocol).                                                                                                                                                                                                                                                                                                                                                                                  |
```

**Review by alai97 — CHANGES_REQUESTED** [review]

File: content/en/database_monitoring/setup_postgres/troubleshooting.md (hunk 3 above)
Comment (djova) [reply content/en/database_monitoring/setup_postgres/troubleshooting.md]: Good question. I couldn't find a built-in sql formatter in `asyncpg`. We should still present a workaround though. I added some wording referring to the `psycopg` formatter.

**Review by alexandre-normand — APPROVED** [review]

**Review by alai97 — APPROVED** [review]

**Pull request merged**